components/service/glean/src/test/java/mozilla/components/service/glean/ping/PingMakerTest.kt

```diff
@@ -93,4 +93,13 @@ class PingMakerTest {
         assertEquals(engine2Data, metricsData.getJSONArray("engine2"))
         assertFalse(metricsData.has("wontCollect"))
     }
+
+    @Test
+    fun `collect() must report an empty string when no data is stored`() {
+        val maker = PingMaker(StorageEngineManager(emptyMap(), applicationContext = mockApplicationContext))
```

Comment:
I think passing in `emptyMap()` here is kind of cheating.  In the normal case, there will be metric types here, but their storage engines will be empty.  I think in that case, we'll still be getting content from the `StorageEngineManager` that looks like:

```
{ "metrics": { "string": {} } }
```

(This is just from looking at the code -- I'm going to confirm my hunch momentarily, but wanted you to get the review asap)...

Reply:
So, I'm happily mistaken, but it's subtle.  If you add a key/value pair to a JSONObject, where the value is `null`, nothing ends up being added.

So, I think I'd be happy with (a) removing the `emptyMap()` argument on this line, and (b) adding a note in StorageEngineManager.collect that we need to be careful to return a completely empty (zero length) JSONObject in order to ensure that pings without any real metrics aren't sent.

Reply:
This is a little trickier than it appears since the GenericScalarStorageEngine.userLifetimeStorage is lazily instantiated.  Without assigning a map here, empty or otherwise, it's provoking a null reference exception when it tries to getSnapshot() since the userLifetimeStorage isn't instantiated yet.  I'm currently looking for a good way around this.

Reply:
Interesting, I'm not seeing that locally, but it could be a test ordering (i.e. "I'm getting lucky") issue.

Reply:
I'm just running this test all by it's lonesome.  Let me try letting them run all in order.

Reply:
I have refactored and pushed a slight change to have data in the map and just request a store that doesn't exist (i.e. it was disabled or nothing was recorded for it).  I hope this covers a little better what your concern is.

Reply:
Yes, indeed that does address my concerns.  As long as we have *some* storage engines in the manager, we're confirming that we don't add the results of empty storage engines.